os/bluestore: kvdb histogram

varadakari opened the pull request:
Generates a histogram of key value distribution for kvdb(rocksdb for now)

ifed01 commented on src/kv/RocksDBStore.h:
```diff
@@ -313,6 +313,8 @@ class RocksDBStore : public KeyValueDB {
     bufferlist value();
     bufferptr value_as_ptr();
     int status();
+    size_t key_size();
```
add override keyword?

ifed01 commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1770,6 +1770,7 @@ class BlueStore : public ObjectStore,
   }
 
   void get_db_statistics(Formatter *f);
+  void get_db_histogram(Formatter *f);
```
'override' here?

ifed01 commented on src/kv/RocksDBStore.h:
```diff
@@ -150,6 +150,8 @@ class RocksDBStore : public KeyValueDB {
   void close();
 
+  void split(const std::string &s, char delim, std::vector<std::string> &elems);
+  std::vector<std::string> split(const std::string &s, char delim);
```
do we really need this form of split?

ifed01 commented on src/kv/RocksDBStore.h:
```diff
@@ -150,6 +150,8 @@ class RocksDBStore : public KeyValueDB {
 
   void close();
 
+  void split(const std::string &s, char delim, std::vector<std::string> &elems);
```
Suggest to rename to split_stats or something to improve readability.

varadakari commented:
@ifed01 fixed the review comments, please review again.

varadakari commented:
sample output of the histogram is below:

{
    "prefix": "B",
    "key-hist": {
        "[0,32)": 208,
        "max_len": 17,
        "value-hist": {
            "[0,64)": 208,
            "max_len": 16
        }   
    },  
    "prefix": "C",
    "key-hist": {
        "[0,32)": 9,
        "max_len": 10,
        "value-hist": {
            "[0,64)": 9,
            "max_len": 10
        }   
    },  
    "prefix": "M",
    "key-hist": {
        "[0,32)": 47,
        "max_len": 26,
        "value-hist": {
            "[0,64)": 31,
            "max_len": 25,
            "[128,192)": 8,
            "max_len": 186,
            "[832,896)": 8,
            "max_len": 847
        },  
        "[32,64)": 12807,
        "max_len": 42,
        "value-hist": {
            "[128,192)": 12807,
            "max_len": 182
        }   
    },

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
+  //histogram
+  struct value_dist
+  {
+    uint64_t count;
```
suggest to initialize struct members with 0 here and below.

varadakari replied:
@ifed01 these are part of the histogram, don't need to specifically initialize to 0, if the value doesn't exist in the map, it will be initialized to 0 by default.

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
+    uint64_t count;
+    uint32_t max_len;
+  };
+  typedef struct value_dist value_dist_t;
```
IMO one can simply define struct value_dist_t above - no need for duplicate type definitions(value_dist & value_dist_t) here.

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
+    uint32_t max_len;
+    map<int, value_dist_t> val_map; // slab id to count and others
+  };
+  typedef struct key_dist key_dist_t;
```
The same here

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
+    pair<string,string> key(iter->raw_key());
+
+    if (key.first == PREFIX_SUPER) {
+	key_hist[PREFIX_SUPER][key_slab].count++;
```
Suggest to introduce a helper func  to update a specific key_hist entry in a single call.
e.g.
update_entry( key_hist, PREFIX_SUPER, key_size, value_size);

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9015,7 +9015,210 @@ int BlueStore::_split_collection(TransContext *txc,
+    }
+    else if (key.first == PREFIX_OBJ) {
+      if (key.second.back() == ONODE_KEY_SUFFIX) {
+	key_hist["o"][key_slab].count++;
```
Suggest to use symbolic names instead of explicit literals: "o", "x", etc

varadakari commented:
@ifed01 addressed the comments, please review again. Have to move the local variables(maps for the histogram) to make the updates as a function. Otherwise rest of the logic is same.

varadakari commented:
@ifed01 Thanks for review, fixed the styling of if-else.

varadakari commented:
i am not hitting this compilation error in my local copy. will rebase and try it again.

varadakari commented:
Fixed all the issues and all tests are passing now.

varadakari commented:
jenkins retest this please

varadakari commented:
@liewegas could you please take look at this?

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9121,7 +9121,182 @@ int BlueStore::_split_collection(TransContext *txc,
 
+const string PREFIX_ONODE = "o";
+const string PREFIX_ONODE_SHARD = "x";
+const string PREFIX_OTHER = "Z";
```
Perhaps instead of #defineing more of these we should just use strings ("onode", "other", etc.)?

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9121,7 +9121,182 @@ int BlueStore::_split_collection(TransContext *txc,
+};
 
+map<string, map<int, struct key_dist> > key_hist;
+map<int, uint64_t> value_hist; // only value distribution histogram
```
Can we put the histogram bits into a class? Should be able to include update_hist_entry?

liewegas commented on src/osd/OSD.cc:
```diff
@@ -1939,6 +1939,8 @@ bool OSD::asok_command(string command, cmdmap_t& cmdmap, string format,
     store->get_db_statistics(f);
   } else if (command == "dump_scrubs") {
     service.dumps_scrub(f);
+  } else if (command == "dump_objectstore_db_histogram") {
```
It might be a good idea to rename this something like "calculate_objectstore_db_histogram" to better suggest/convey that this is a very slow operation?

varadakari commented:
@liewegas Made the changes and repushed after rebasing, could you please review again?

varadakari commented:
jenkins retest this please

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9118,7 +9118,164 @@ int BlueStore::_split_collection(TransContext *txc,
+  f->dump_unsigned("total_value_size", total_value_size);
+  f->close_section();
+
+  f->open_object_section("rocksdb value distribution");
```
no spaces in formatter fields names please

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9118,7 +9118,164 @@ int BlueStore::_split_collection(TransContext *txc,
+  f->open_object_section("rocksdb key-value histogram");
+  for (auto i : hist.key_hist) {
+    f->dump_string("prefix", i.first);
+    f->open_object_section("key-hist");
```
_ not -

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9118,7 +9118,164 @@ int BlueStore::_split_collection(TransContext *txc,
+  f->close_section();
+
+  f->open_object_section("rocksdb key-value histogram");
+  for (auto i : hist.key_hist) {
```
the dump of the histogram itself should be DBHistogram::dump()

liewegas commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1496,6 +1496,30 @@ class BlueStore : public ObjectStore,
 
+  struct DBHistogram {
+    struct value_dist
+    {
```
{ on previous line

liewegas commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1496,6 +1496,30 @@ class BlueStore : public ObjectStore,
+    {
+      uint64_t count;
+      uint32_t max_len;
+      map<int, struct value_dist> val_map; // slab id to count and others
```
these comments should be like ///< so the doc generation can do its thing

varadakari commented:
@liewegas Addressed the comments. Please have relook at the changes.

liewegas commented:
Can you add a test to store_test.cc that triggers this so that the code path is tested?

liewegas commented:
@varadakari ping

varadakari commented:
@liewegas Added the unit test cases for stats and histogram. Please have a look.

ceph-jenkins commented:
submodules for project are unmodified

varadakari commented:
@liewegas if the tests are complete, could you merge this please?
